x/contracts/examples/multisig/src/lib.rs

```diff
@@ -0,0 +1,468 @@
+    ProposalItem(ProposalId) => Proposal,
+    ProposalVoteInfo(ProposalId) => ProposalMeta,
+    Vote(ProposalId, Address) => bool,
+}
+
+#[public]
+pub fn propose(ctx: &mut Context, proposal: Proposal, meta: ProposalMeta) -> ProposalId {
+    let actor = ctx.actor();
+
+    let id = ctx
+        .get(ProposalCount)
+        .expect("failed to deserialize")
+        .unwrap_or_default();
+
+    ctx.store((
+        (ProposalCount, id + 1),
+        (ProposalItem(id), proposal),
+        (ProposalVoteInfo(id), meta),
+        (Vote(id, actor), true),
+    ))
+    .expect("failed to store proposal");
+
+    id
+}
+
+#[public]
+pub fn vote(ctx: &mut Context, proposal_id: ProposalId, vote: bool) -> Option<Vec<u8>> {
+    let actor = ctx.actor();
+
+    ctx.store_by_key(Vote(proposal_id, actor), vote)
+        .expect("failed to serialize");
+
+    if !vote {
+        return None;
+    }
+
+    let ProposalMeta {
+        voters,
+        votes_remaining,
+    } = ctx
+        .get(ProposalVoteInfo(proposal_id))
+        .expect("failed to deserialize")
+        .expect("proposal not found");
+
+    if !voters.contains(actor) || votes_remaining == 0 {
+        return None;
+    }
+
+    let votes_remaining = votes_remaining - 1;
+
+    ctx.store_by_key(
+        ProposalVoteInfo(proposal_id),
+        ProposalMeta {
+            voters,
+            votes_remaining,
+        },
+    )
+    .expect("failed to store votes remaining");
+
+    match votes_remaining {
+        0 => {
+            let Proposal {
+                contract_address,
+                function_name,
+                args,
+            } = ctx
+                .get(ProposalItem(proposal_id))
+                .expect("failed to deserialize")
+                .expect("proposal not found");
+            // TODO: fix this
+            // let remaining_fuel = ctx.remaining_fuel();
+            let remaining_fuel = 0;
+
+            let result: DeferDeserialization = ctx
+                .call_contract(contract_address, &function_name, &args, remaining_fuel, 0)
+                .expect("failed to call contract");
+
+            Some(result.into())
+        }
+
+        _ => None,
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+#[borsh(crate = "wasmlanche::borsh")]
+pub struct Proposal {
+    contract_address: Address,
+    function_name: String,
+    args: Vec<u8>,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+#[borsh(crate = "wasmlanche::borsh")]
+pub struct ProposalMeta {
+    voters: Voters,
+    votes_remaining: usize,
+    // TODO: add expiry
+}
+
+#[cfg_attr(test, derive(Debug))]
+pub enum Voters {
+    Any,
+    Addresses(HashSet<Address>),
+}
+
+impl Voters {
+    pub fn with_quorum(self, quorum: NonZeroUsize) -> Result<ProposalMeta, Self> {
+        let quorum = quorum.get();
+
+        // When one makes a proposal, they automatically vote yes for it
+        if quorum > 1 && quorum <= self.len() {
+            let votes_remaining = quorum - 1;
```

Comment:
As you vote "yes" for the proposal maker, you might as well directly remove it from the set of voters or it will be able to double-vote (if you go for the removal of voters route to avoid this issue).

Reply:
Good idea

Reply:
Closing this since the proposer casts a vote and a call to the vote function returns `None` and doesn't change the state if the actor already voted for this proposal.